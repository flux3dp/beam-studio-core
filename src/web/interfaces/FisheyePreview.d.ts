import { IDeviceInfo } from './IDevice';

export interface FisheyeMatrix {
  k: number[][];
  d: number[][];
  points: [number, number][][];
  center?: [number, number];
}

export interface FisheyeCameraParametersV1 {
  k: number[][];
  d: number[][];
  heights: number[];
  points: [number, number][][][];
  center: [number, number];
  z3regParam: number[][][][]; // [i][j][k][l] i: row, j: column, k: x/y, l: 3/2/1/0 th order;
}

export interface FisheyeCameraParametersV2Cali {
  refHeight?: number;
  k?: number[][];
  d?: number[][];
  rvec?: number[];
  tvec?: number[];
  points?: [number, number][][];
  // 3rd order regression parameters for camera position with respect to y position
  hx?: number[];
  hy?: number[];
  xc?: number[];
  yc?: number[];
  imageScale?: number[];
  dh?: number;
}

export interface FisheyeCameraParametersV2 {
  refHeight: number;
  k: number[][];
  d: number[][];
  rvec: number[];
  tvec: number[];
  points: [number, number][][];
  // 3rd order regression parameters for camera position with respect to y position
  hx: number[];
  hy: number[];
  xc: number[];
  yc: number[];
  imageScale: number[];
  v: 2;
}

export type FisheyeCameraParameters = FisheyeCameraParametersV1 | FisheyeCameraParametersV2;

/**
 * RotationParameters3D to save in the machine
 */
export interface RotationParameters3D {
  // rotation in 3 axes
  rx: number;
  ry: number;
  rz: number;
  // sh: Scale of h, since the dimension of the image is in pixel, the height when previewing is in mm,
  // we need the scale of h to convert the height in mm to pixel, the value is usually 6
  sh: number;
  // ch: constant of h, the distant from the top position of probe to the camera,
  // the value is usually 162mm in mechanical
  ch: number;
  // dh: the height deviation of the camera, would apply to preview height
  dh: number;
}

/**
 * RotationParameters3D for calibration
 * add tx and ty to handle translation error
 * tx and ty will be saved in FisheyeMatrix center
 */
export interface RotationParameters3DCalibration extends RotationParameters3D {
  tx: number;
  ty: number;
}

/**
 * RotationParameters3D for ghost api,
 * need to calculate h from sh, ch and object height
 */
export interface RotationParameters3DGhostApi {
  // rotation in 3 axes
  rx: number;
  ry: number;
  rz: number;
  h: number;
  tx: number;
  ty: number;
}

interface FisheyePreviewManagerBase {
  version: number;
  device: IDeviceInfo;
  params: FisheyeCameraParameters;
  objectHeight: number;
  levelingOffset: Record<string, number>;

  setupFisheyePreview(progressId?: string): Promise<boolean>;

  onObjectHeightChanged(): Promise<void>;

  resetObjectHeight(): Promise<boolean>;

  reloadLevelingOffset(): Promise<void>;
}

export interface FisheyePreviewManager {
  version: number;
  device: IDeviceInfo;
  params: FisheyeCameraParameters;
  support3dRotation: boolean;
  rotationData?: RotationParameters3DCalibration;
  objectHeight: number;
  levelingData: Record<string, number>;
  levelingOffset: Record<string, number>;

  setupFisheyePreview(progressId?: string): Promise<boolean>;

  update3DRotation?(newData: RotationParameters3DCalibration): Promise<void>;

  onObjectHeightChanged(): Promise<void>;

  resetObjectHeight(): Promise<boolean>;

  reloadLevelingOffset(): Promise<void>;
}
