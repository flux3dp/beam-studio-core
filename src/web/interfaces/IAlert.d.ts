/* eslint-disable @typescript-eslint/ban-types */
import { IButton } from './IButton';

export interface IAlert {
  id?: string;
  type?: string;
  message?: string;
  caption?: string;
  iconUrl?: string;
  children?: Element;
  buttons?: IButton | IButton[];
  buttonType?: string;
  buttonLabels?: string | string[];
  callbacks?: Function | Function[];
  primaryButtonIndex?: number;
  onYes?: Function;
  onNo?: Function;
  onConfirm?: Function;
  onRetry?: Function;
  onCancel?: Function;
  checkbox?: ICheckbox;
  checkboxText?: string;
  checkboxCallbacks?: Function | Function[];
}

export interface ICheckbox {
  text: string;
  callbacks: Function | Function[];
  onYes?: Function;
  onNo?: Function;
  onConfirm?: Function;
  onRetry?: Function;
  onCancel?: Function;
}
