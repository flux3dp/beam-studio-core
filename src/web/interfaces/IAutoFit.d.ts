export interface AutoFit {
  angle: number;
  bbox: number[];
  center: number[];
}
