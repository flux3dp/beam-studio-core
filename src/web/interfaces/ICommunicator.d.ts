export interface ICommunicator {
  on(channel: string, listener: any): void;
  send(channel: string, ...args: any[]): void;
}
