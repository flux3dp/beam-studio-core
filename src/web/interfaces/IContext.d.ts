export interface ILayerPanelContext {
  selectedLayers?: string[],
  updateLayerPanel: () => void,
  setSelectedLayers: (layers: string[]) => null,
}
