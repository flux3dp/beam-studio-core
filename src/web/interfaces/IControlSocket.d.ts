import EventEmitter from 'eventemitter3';

import { SwiftrayClient } from 'helpers/api/swiftray-client';

// eslint-disable-next-line import/no-cycle
import { IDeviceDetailInfo, IReport } from './IDevice';
import { WrappedWebSocket } from './WebSocket';
import { RawChipSettings } from './Cartridge';

interface IControlSocket extends EventEmitter {
  isConnected: boolean;
  connection: WrappedWebSocket | SwiftrayClient | null;
  isLineCheckMode: boolean;
  lineNumber: number;

  getMode(): string;
  // eslint-disable-next-line @typescript-eslint/ban-types
  addTask<T>(taskFunction: (...args) => T, ...args: unknown[]): Promise<T>;
  connect(): Promise<void>;
  killSelf(): Promise<null>;
  setProgressListener(listener: (...args: unknown[]) => void): void;
  removeCommandListeners(): void;

  ls(path: string): Promise<{ files: string[]; directories: string[] }>;
  lsusb(): Promise<unknown>;
  fileInfo(path: string, fileName: string): Promise<unknown[]>;
  report(): Promise<{ device_status: IReport }>;
  upload(data: any, path?: string, fileName?: string): Promise<void>;
  abort(): Promise<unknown>;
  quit(): Promise<unknown>;
  start(): Promise<unknown>;
  pause(): Promise<unknown>;
  resume(): Promise<unknown>;
  restart(): Promise<unknown>;
  kick(): Promise<unknown>;
  quitTask(): Promise<unknown>;
  deviceDetailInfo(): Promise<IDeviceDetailInfo>;
  getPreview(): Promise<unknown[]>;
  select(path: string[], fileName: string): Promise<{ status: string }>;
  deleteFile(fileNameWithPath: string): Promise<unknown>;
  downloadFile(fileNameWithPath: string): Promise<(string | Blob)[]>;
  downloadLog(logName: string): Promise<(string | Blob)[]>;
  getLaserPower(): Promise<{ value: number }>;
  setLaserPower(power: number): Promise<unknown>;
  setLaserPowerTemp(power: number): Promise<unknown>;
  getLaserSpeed(): Promise<{ value: number }>;
  setLaserSpeed(speed: number): Promise<unknown>;
  setLaserSpeedTemp(speed: number): Promise<unknown>;
  getFan(): Promise<{ value: number }>;
  setFan(fanSpeed: number): Promise<unknown>;
  setFanTemp(fanSpeed: number): Promise<unknown>;
  setOriginX(x: number): Promise<unknown>;
  setOriginY(y: number): Promise<unknown>;
  getDoorOpen(): Promise<{ value: string }>;
  getDeviceSetting(name: string): Promise<unknown>;
  setDeviceSetting(name: string, value: string): Promise<unknown>;
  deleteDeviceSetting(name: string): Promise<unknown>;
  enterMaintainMode(): Promise<unknown>;
  endMaintainMode(): Promise<unknown>;
  maintainMove(args: { x?: number; y?: number; z?: number; f?: number }): Promise<unknown>;
  maintainCloseFan(): Promise<unknown>;
  maintainHome(): Promise<unknown>;
  enterRawMode(): Promise<unknown>;
  endRawMode(): Promise<unknown>;
  rawHome(zAxis?: boolean): Promise<void>;
  rawMoveZRelToLastHome(z: number): Promise<unknown>;
  rawStartLineCheckMode(): Promise<void>;
  rawEndLineCheckMode(): Promise<void>;
  rawMove(args: { x?: number; y?: number; z?: number; f?: number }): Promise<unknown>;
  rawSetWaterPump(on: boolean): Promise<unknown>;
  rawSetAirPump(on: boolean): Promise<unknown>;
  rawSetFan(on: boolean): Promise<unknown>;
  rawSetRotary(on: boolean): Promise<unknown>;
  rawSetLaser(args: { on: boolean; s?: number }): Promise<unknown>;
  rawSet24V(on: boolean): Promise<unknown>;
  rawAutoFocus(timeout?: number): Promise<void>;
  fwUpdate(file: File): Promise<unknown>;

  // method not supported by SwiftrayClient
  fetchCameraCalibImage?: (name?: string) => Promise<Blob>;
  enterCartridgeIOMode?: () => Promise<void>;
  endCartridgeIOMode?: () => Promise<void>;
  getCartridgeChipData?: () => Promise<{ status: string; data: { result: RawChipSettings } }>;
  cartridgeIOJsonRpcReq?: (
    method: string,
    params: unknown
  ) => Promise<{ status: string; data: { result: { hash: string; sign: string } } }>;
}

export default IControlSocket;
