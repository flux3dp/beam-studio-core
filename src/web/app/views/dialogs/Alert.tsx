/**
 * new Alert Modal using antd Modal
 */
import classNames from 'classnames';
import React, { useContext, useState } from 'react';
import { Button, Checkbox, Modal } from 'antd';

import AlertIcons from 'app/icons/alerts/AlertIcons';
import { AlertProgressContext } from 'app/contexts/AlertProgressContext';
import { IAlert } from 'interfaces/IAlert';

import styles from './Alert.module.scss';

const renderIcon = (url?: string): JSX.Element => {
  if (!url) return null;
  return <img className={styles.icon} src={url} />;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const messageIconMap: { [key: string]: (props: any) => JSX.Element } = {
  notice: AlertIcons.Notice,
};

const renderMessage = (message: string | React.ReactNode, messageIcon = ''): JSX.Element => {
  if (!message) return null;
  let content = null;
  const IconComponent = messageIconMap[messageIcon];
  if (typeof message === 'string') {
    content = (
      <div
        className={classNames(styles.message, { [styles['with-icon']]: !!IconComponent })}
        // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{ __html: message }}
      />
    );
  } else {
    content = (
      <div className={classNames(styles.message, { [styles['with-icon']]: !!IconComponent })}>
        {message}
      </div>
    );
  }
  return (
    <div className={styles['message-container']}>
      {IconComponent && <IconComponent className={styles.icon} />}
      {content}
    </div>
  );
};

interface Props {
  data: IAlert;
}

const Alert = ({ data }: Props): JSX.Element => {
  const { popFromStack } = useContext(AlertProgressContext);
  const { caption, checkbox, message, messageIcon, buttons, iconUrl } = data;

  const [checkboxChecked, setCheckboxChecked] = useState(false);

  const renderCheckbox = (): JSX.Element => {
    if (!checkbox) return null;
    const { text } = checkbox;
    return <Checkbox onClick={() => setCheckboxChecked(!checkboxChecked)}>{text}</Checkbox>;
  };

  const footer = buttons?.map((button, idx) => {
    const buttonType = button.className?.includes('primary') ? 'primary' : 'default';
    return (
      <Button
        key={button.label}
        onClick={() => {
          popFromStack();
          if (checkbox && checkboxChecked) {
            const { callbacks } = checkbox;
            if (callbacks.length > idx) callbacks[idx]?.();
            else if (typeof callbacks === 'function') callbacks?.();
            else button.onClick?.();
          } else button.onClick?.();
        }}
        type={buttonType}
      >
        {button.label}
      </Button>
    );
  });

  return (
    <Modal
      open
      title={caption}
      footer={footer}
      closable={false}
      maskClosable={false}
      centered
      onCancel={popFromStack}
    >
      {renderIcon(iconUrl)}
      {renderMessage(message, messageIcon)}
      {renderCheckbox()}
    </Modal>
  );
};

export default Alert;
