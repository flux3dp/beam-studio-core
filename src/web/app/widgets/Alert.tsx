import * as React from 'react';
import classNames from 'classnames';

import browser from 'implementations/browser';
import ButtonGroup from 'app/widgets/Button-Group';
import Modal from 'app/widgets/Modal';

interface Props {
  caption: string;
  iconUrl?: string;
  message: string | JSX.Element;
  checkboxText?: string;
  checkboxCallbacks?: () => void | (() => void)[];
  onClose?: () => void;
  animationClass?: string;
  buttons: {
    label: string;
    className: string;
    onClick: () => void;
  }[];
}

interface State {
  checkboxChecked: boolean;
}

class Alert extends React.Component<Props, State> {
  private messageRef: React.RefObject<HTMLPreElement>;

  constructor(props: Props) {
    super(props);
    this.state = {
      checkboxChecked: false,
    };
    this.messageRef = React.createRef();
  }

  componentDidMount(): void {
    const message = this.messageRef.current as Element;
    if (message) {
      const aElements = message.querySelectorAll('a');
      for (let i = 0; i < aElements.length; i += 1) {
        const a = aElements[i];
        a.addEventListener('click', (e) => {
          e.preventDefault();
          browser.open(a.getAttribute('href'));
        });
      }
    }
  }

  renderCaption = (): JSX.Element => {
    const { caption } = this.props;
    if (!caption) return null;

    return (
      <h2 className="caption">{caption}</h2>
    );
  };

  renderIcon = (): JSX.Element => {
    const { iconUrl } = this.props;
    if (!iconUrl) return null;

    return (
      <img className="icon" src={iconUrl} />
    );
  };

  renderMessage = (): JSX.Element => {
    const { message } = this.props;
    return typeof message === 'string'
      // eslint-disable-next-line react/no-danger
      ? <pre ref={this.messageRef} className="message" dangerouslySetInnerHTML={{ __html: message }} />
      : <pre className="message">{message}</pre>;
  };

  renderCheckbox = (): JSX.Element => {
    const { checkboxText } = this.props;
    const { checkboxChecked } = this.state;
    if (!checkboxText) return null;

    return (
      <div className="modal-checkbox">
        <input
          type="checkbox"
          onClick={() => this.setState({ checkboxChecked: !checkboxChecked })}
        />
        {checkboxText}
      </div>
    );
  };

  renderChildren = (): JSX.Element => {
    const { children } = this.props;
    if (!children) return null;

    return (
      <div className="alert-children">
        {children}
      </div>
    );
  };

  render = (): JSX.Element => {
    const {
      checkboxText, checkboxCallbacks, onClose, animationClass,
    } = this.props;
    let { buttons } = this.props;
    const { checkboxChecked } = this.state;
    buttons = buttons.map((b, i) => {
      const newButton = { ...b };
      const buttonCallback = b.onClick;
      if (!checkboxChecked || !checkboxText || !checkboxCallbacks) {
        newButton.onClick = () => {
          if (onClose) onClose();
          buttonCallback();
        };
      } else if (typeof checkboxCallbacks === 'function') {
        // Need to reset checkbox state after callback
        newButton.onClick = () => {
          // If only one checkbox callback passed, run checkbox callback after
          // runing button callback
          if (onClose) onClose();
          buttonCallback();
          checkboxCallbacks();
        };
      } else if ((checkboxCallbacks as (() => void)[]).length > i) {
        newButton.onClick = () => {
          // If more than one checkbox callbacks passed,
          // replace original checkbox callbacks.
          if (onClose) onClose();
          (checkboxCallbacks as (() => void)[])[i]();
        };
      } else {
        newButton.onClick = () => {
          if (onClose) onClose();
          buttonCallback();
        };
      }
      return newButton;
    });

    return (
      <Modal>
        <div className={classNames('modal-alert', animationClass)}>
          {this.renderCaption()}
          {this.renderIcon()}
          {this.renderMessage()}
          {this.renderChildren()}
          {this.renderCheckbox()}
          <ButtonGroup buttons={buttons} />
        </div>
      </Modal>
    );
  };
}

export default Alert;
