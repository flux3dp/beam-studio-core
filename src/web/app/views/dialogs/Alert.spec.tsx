import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import { IAlert } from 'interfaces/IAlert';

import Alert from './Alert';

const mockPopFromStack = jest.fn();
jest.mock('app/contexts/AlertProgressContext', () => ({
  AlertProgressContext: React.createContext({
    popFromStack: () => mockPopFromStack,
  }),
}));

const mockOnYes = jest.fn();
const mockOnNo = jest.fn();
const mockOnCheckedYes = jest.fn();
const mockOnCheckedNo = jest.fn();

const mockData: IAlert = {
  id: 'alert',
  message: 'Yes or No',
  caption: 'Hello World',
  iconUrl: 'https://www.flux3dp.com/icon.svg',
  buttons: [{
    title: 'Yes',
    label: 'Yes',
    onClick: mockOnYes,
  }, {
    title: 'No',
    label: 'No',
    onClick: mockOnNo,
  }],
  checkbox: {
    text: 'checkbox',
    callbacks: [mockOnCheckedYes, mockOnCheckedNo],
  }
};

describe('test Alert', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { baseElement } = render(<Alert data={mockData} />);
    expect(baseElement).toMatchSnapshot();
  });

  test('should call callback when click button', () => {
    const { getByText } = render(<Alert data={mockData} />);
    fireEvent.click(getByText('Yes'));
    expect(mockOnYes).toBeCalledTimes(1);
    fireEvent.click(getByText('No'));
    expect(mockOnNo).toBeCalledTimes(1);
    fireEvent.click(getByText('checkbox'));

    expect(mockOnCheckedYes).not.toBeCalled();
    fireEvent.click(getByText('Yes'));
    expect(mockOnCheckedYes).toBeCalledTimes(1);
    expect(mockOnYes).toBeCalledTimes(1);

    expect(mockOnCheckedNo).not.toBeCalled();
    fireEvent.click(getByText('No'));
    expect(mockOnCheckedNo).toBeCalledTimes(1);
    expect(mockOnNo).toBeCalledTimes(1);
  });
});
