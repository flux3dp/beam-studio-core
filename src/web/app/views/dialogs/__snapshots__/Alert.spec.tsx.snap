// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Alert should render correctly 1`] = `
<body>
  <div />
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-hash"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap ant-modal-centered"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal css-dev-only-do-not-override-hash ant-zoom-appear ant-zoom-appear-prepare ant-zoom"
          role="dialog"
          style="width: 520px;"
        >
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
          <div
            class="ant-modal-content"
          >
            <div
              class="ant-modal-header"
            >
              <div
                class="ant-modal-title"
                id="test-id"
              >
                Hello World
              </div>
            </div>
            <div
              class="ant-modal-body"
            >
              <img
                class="icon"
                src="https://www.flux3dp.com/icon.svg"
              />
              <div
                class="message-container"
              >
                <div
                  class="message"
                >
                  Yes or No
                </div>
              </div>
              <label
                class="ant-checkbox-wrapper css-dev-only-do-not-override-hash"
              >
                <span
                  class="ant-checkbox ant-wave-target css-dev-only-do-not-override-hash"
                >
                  <input
                    class="ant-checkbox-input"
                    type="checkbox"
                  />
                  <span
                    class="ant-checkbox-inner"
                  />
                </span>
                <span>
                  checkbox
                </span>
              </label>
            </div>
            <div
              class="ant-modal-footer"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-hash ant-btn-default"
                type="button"
              >
                <span>
                  Yes
                </span>
              </button>
              <button
                class="ant-btn css-dev-only-do-not-override-hash ant-btn-default"
                type="button"
              >
                <span>
                  No
                </span>
              </button>
            </div>
          </div>
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`test Alert should render correctly with help center link 1`] = `
<body>
  <div />
  <div>
    <div
      class="ant-modal-root css-dev-only-do-not-override-hash"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap ant-modal-centered"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal css-dev-only-do-not-override-hash ant-zoom-appear ant-zoom-appear-prepare ant-zoom"
          role="dialog"
          style="width: 520px;"
        >
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
          <div
            class="ant-modal-content"
          >
            <div
              class="ant-modal-header"
            >
              <div
                class="ant-modal-title"
                id="test-id"
              >
                Hello World
              </div>
            </div>
            <div
              class="ant-modal-body"
            >
              <img
                class="icon"
                src="https://www.flux3dp.com/icon.svg"
              />
              <div
                class="message-container"
              >
                <div
                  class="message"
                >
                  #801 error
                </div>
              </div>
              <button
                class="ant-btn css-dev-only-do-not-override-hash ant-btn-link link"
                type="button"
              >
                <span>
                  Learn more
                </span>
              </button>
              <label
                class="ant-checkbox-wrapper css-dev-only-do-not-override-hash"
              >
                <span
                  class="ant-checkbox ant-wave-target css-dev-only-do-not-override-hash"
                >
                  <input
                    class="ant-checkbox-input"
                    type="checkbox"
                  />
                  <span
                    class="ant-checkbox-inner"
                  />
                </span>
                <span>
                  checkbox
                </span>
              </label>
            </div>
            <div
              class="ant-modal-footer"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-hash ant-btn-default"
                type="button"
              >
                <span>
                  Yes
                </span>
              </button>
              <button
                class="ant-btn css-dev-only-do-not-override-hash ant-btn-default"
                type="button"
              >
                <span>
                  No
                </span>
              </button>
            </div>
          </div>
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
